python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -896,6 +896,114 @@ def corr(self, method="pearson", min_periods=1):
 
         return res
 
+    def cov(self, min_periods=None, ddof=1):
+        """
+        Compute pairwise covariance of columns, excluding NA/null values.
+
+        Parameters
+        ----------
+        min_periods: int, optional
+            Minimum number of observations required per pair of columns
+            to have a valid result.
+
+        ddof: int, optional
+            Delta degrees of freedom, default is 1.
+
+        Returns
+        ----------
+        DataFrame
+            Covariance matrix.
+
+        Examples
+        --------
+        >>> import cudf
+        >>> gdf = cudf.DataFrame({
+        ...             "id": ["a", "a", "a", "b", "b", "b", "c", "c", "c"],
+        ...             "val1": [5, 4, 6, 4, 8, 7, 4, 5, 2],
+        ...             "val2": [4, 5, 6, 1, 2, 9, 8, 5, 1],
+        ...             "val3": [4, 5, 6, 1, 2, 9, 8, 5, 1]})
+        >>> gdf
+        id  val1  val2  val3
+        0  a     5     4     4
+        1  a     4     5     5
+        2  a     6     6     6
+        3  b     4     1     1
+        4  b     8     2     2
+        5  b     7     9     9
+        6  c     4     8     8
+        7  c     5     5     5
+        8  c     2     1     1
+        >>> gdf.groupby("id").cov()
+                    val1       val2       val3
+        id
+        a  val1  1.000000   0.500000   0.500000
+           val2  0.500000   1.000000   1.000000
+           val3  0.500000   1.000000   1.000000
+        b  val1  4.333333   3.500000   3.500000
+           val2  3.500000  19.000000  19.000000
+           val3  3.500000  19.000000  19.000000
+        c  val1  2.333333   3.833333   3.833333
+           val2  3.833333  12.333333  12.333333
+           val3  3.833333  12.333333  12.333333
+        """
+
+        # create expanded dataframe consisting all combinations of the
+        # struct columns-pairs to be correlated
+        # i.e (('col1', 'col1'), ('col1', 'col2'), ('col2', 'col2'))
+        _cols = self.grouping.values.columns.tolist()
+        len_cols = len(_cols)
+
+        new_df_data = {}
+        for x, y in itertools.combinations_with_replacement(_cols, 2):
+            new_df_data[(x, y)] = cudf.DataFrame._from_data(
+                {"x": self.obj._data[x], "y": self.obj._data[y]}
+            ).to_struct()
```

Comment:
Does this use copies or (non-owning) views of the existing columns?

Reply:
`to_struct` will make a deep copy of the data.

Reply:
Oof. Making deep copies for N^2 columns would consume a large amount of GPU memory. We only need a view here. If we need to materialize a deep copy, we surely shouldn’t do it for all column pairs at once.

Reply:
I ran a quick test of this with @mayankanand007. Using a dataframe with 100 columns and 30,000 rows, we run out of memory (48 GB V100). This seems like a small problem size for a covariance matrix in cuDF -- we should definitely refactor to use views here, and it's on the edge of what I would consider a blocker for this feature.

Reply:
Making an issue for `to_struct`.

Reply:
> my suggestion is to deprecate the method and move to internal namespace _to_struct, add a new parameter copy for optimization uses. (Out of scope of this PR for sure.)

Common big data processing engines like Spark, Hive, etc. allow users to create structs from existing columns (by copy) and it's fairly common. When a user needs to create a struct column (perhaps to interact with another service), it can be challenging to move forward without a constructor in the API.

Perhaps we might consider an internal API that exposes copy vs view semantics as you suggest without deprecating the public API?

Reply:
@beckernick Right on. I discussed this on a call with @isVoid (before your latest comment), and we agreed there is a need to expose a constructor of some kind. It does seem like we need to think about adding view semantics for computing groupby quantities over structs, most likely via some internal APIs.

In my understanding, one catch is that struct columns have specific requirements on their parent/child null masks (pushdown semantics). A method for constructing a "struct column view" (already an overloaded term, I think?) from child column views and a nontrivial parent null mask (one that is not all valid) would probably have to enforce some requirements on the parent/child null masks at the time of construction.

Reply:
It seems like our concerns about view/copy semantics have been addressed by @isVoid's suggestions in this comment https://github.com/rapidsai/cudf/pull/9889#discussion_r784468894 and the changes in commit 2ea064dcbc1b03aaa2034d4de30b850f69eeecd5. Resolving this thread.